fix example solution & add --config option

alexandre-allard opened the pull request:
**Component**: examples

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged

Closes: #ISSUE_NUMBER

-->
<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

alexandre-allard committed:
examples: rename example operator

Rename the operator to match the pattern
<solution-name>-operator:
example-operator -> example-solution-operator

alexandre-allard committed:
examples: add a --config flag to operator

Add an option to pass a configuration to the operator
For the moment, only the configuration for the registry
prefixes are handled.

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/fix-example-solution` into
`development/2.4`
* [w/2.5/bugfix/fix-example-solution](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/fix-example-solution) will be merged into `development/2.5` 
* [w/2.6/bugfix/fix-example-solution](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/fix-example-solution) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #bc1656e62d95e857831aa783cfe74e2543ce2a4c on the integration branch
`w/2.5/bugfix/fix-example-solution` is merging a branch which is neither the current
branch `bugfix/fix-example-solution` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/fix-example-solution` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
Did you try upgrading to the latest Operator SDK?

gdemonet commented on examples/metalk8s-solution-example/operator/go.mod:
```diff
@@ -1,10 +1,11 @@
 	github.com/go-openapi/spec v0.19.0
 	github.com/operator-framework/operator-sdk v0.9.1-0.20190805223000-66e78cc576ef
 	github.com/spf13/pflag v1.0.3
+	gopkg.in/yaml.v2 v2.2.2
```
Did you look into [`spf13/viper`](https://github.com/spf13/viper)?

gdemonet replied:
It supports a bunch of features out-of-the-box, such as live reloading of config files, that would help us a lot.

alexandre-allard replied:
No, didn't check this, I will.

NicolasT replied:
Why would you want live-reloading of configs? :-/

gdemonet replied:
Importing a new Solution version would immediately allow Operators to pick it up if we propagate the repositories in their ConfigMaps.

NicolasT replied:
But that operator will not know about said version (or how to deploy it). So a restart would be required anyway.

In general, an app which starts, loads config, logs it and then uses that (static) config until it gets restarted is much easier to troubleshoot than some app which can dynamically at random points in time change its config, IMHO.

Also, operators are (or should be) trivial to restart (that's the whole point of stateless control loops).

gdemonet replied:
OK, then maybe we don't need those extra features.

I was looking at `prometheus-operator`, and they seem to use the standard [`flag`](https://golang.org/pkg/flag/) package: see [here](https://github.com/coreos/prometheus-operator/blob/3a86bb3115d59db459c92fb54ea63248ee4e4d3a/cmd/operator/main.go#L124-L158)
No way to pass a config file here.

`kubectl` uses `spf13/cobra` and some complex patterns to load its kubeconfig (tried to read through the code but it's confusing for a noob like me), registering all config options with command-line flag overrides.

Since Operator SDK sets up the manager with `spf13/pflag`, it's not too far from how `spf13/viper` works (e.g. see [this PR](https://github.com/operator-framework/operator-sdk/pull/950/files) converting from one to the other). Maybe we don't care about live-reload, but having a config-file with command-line flags overrides is nice for testing or simpler setups.
We can build our own in-house parsing of such files, but it seems the ecosystem often relies on `viper` and `cobra` for Go CLIs, so I wouldn't worry too much about it and just follow the trend.

Also, not sure if there's a standard way of validating an `apiVersion`/`kind` tuple in the tooling controllers rely on, but I'd assume there is...

gdemonet replied:
(of course, all those changes can be considered in a future PR! we just need to think about "real" Operators that will be built around this tooling, and the closer we can get to this with the example, the more likely we'll define better tooling)

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,39 @@
+
+type TOperatorConfig struct {
+        Kind         string `yaml:"kind"`
+        APIVersion   string `yaml:"apiVersion"`
```
Those values should be checked before the complete `yaml.Unmarshal`, to ensure we understand the provided format and warn the user properly if we don't.

alexandre-allard replied:
I've added a validator, tell me if you have a better idea in mind.

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/controller/util/baseserver.go:
```diff
@@ -109,9 +110,12 @@ func BuildLabelSelector(labels map[string]string) metav1.LabelSelector {
-	prefix, found := os.LookupEnv("REGISTRY_PREFIX")
-	if !found {
-		prefix = "docker.io/metalk8s"
+	var prefix string = "docker.io/metalk8s"
```
Let's define this as a default value for the config and only read from config in controllers.

See https://github.com/spf13/viper#establishing-defaults

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/controller/util/baseserver.go:
```diff
@@ -109,9 +110,12 @@ func BuildLabelSelector(labels map[string]string) metav1.LabelSelector {
+	for _, registry := range config.OperatorConfig.Repositories {
+		if registry.Version == version {
+			prefix = registry.Repository
+		}
```
Not entirely sure about this logic. Consider a complex Operator which not only deploys custom images (which will have the same version as the Operator, probably), but also some open-sourced components (e.g. a well-known database or queue system). The version of said images would not be the same as the `repository.Version`, yet would live in one specific repository.

Mapping between those would need to live somewhere, and I don't know what would be the most flexible approach here. Thoughts @benoit-a @RachedBenMustapha?

alexandre-allard replied:
Yeah right, didn't though of other Solutions here.
Will try to find a better format for the repository mapping.

rachedbenmustapha replied:
@gdemonet for Zenko Operator, since we have all possible cases you mentioned (including being used outside of the solution framework), we use a combination of an image prefix (not yet populated, always empty, but used in all generated ~workflows~ workloads) and explicit repos/versions in a [ZenkoVersoin CR](https://github.com/scality/zenko-operator/blob/master/doc/examples/zenkoversion-1.1-dev.yaml) that is going to be generated in the release process and shipped inside a version's solution ISO.

alexandre-allard commented:
> Did you try upgrading to the latest Operator SDK?

No, why ?

NicolasT commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,39 @@
+        } `yaml:"repositories"`
+}
+
+var OperatorConfig TOperatorConfig
```
Not sure some global mutable config 'singleton' is a great pattern.

alexandre-allard replied:
I agree, I've changed this a little bit.

gdemonet commented:
> > Did you try upgrading to the latest Operator SDK?
> 
> No, why ?

To see if it changes anything :innocent: we can track this in another ticket, no worries.

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,153 @@
+
+type OperatorConfig struct {
+	Kind         string `yaml:"kind" validate:"required,eq=OperatorConfig"`
+	APIVersion   string `yaml:"apiVersion" validate:"required"`
```
Would be nice to validate we know the APIVersion as well ;)

alexandre-allard replied:
Yes, sure.

alexandre-allard replied:
done

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,153 @@
+	return message
+}
+
+func validate() error {
```
Why not make this as a method on any `OperatorConfig` struct?

alexandre-allard replied:
done

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,153 @@
+		return err
+	}
+
+	return nil
```
I'd expect this method to return the config struct if it could load. Something like:

```go
func LoadConfiguration(input io.Reader) (*OperatorConfig, error)  {
```

So we can easily call it with the usual:

```go
config, err := LoadConfiguration(file)
if err != nil {
    // do something with err
}
// consume config
```

alexandre-allard replied:
done

gdemonet commented on examples/metalk8s-solution-example/operator/pkg/controller/util/baseserver.go:
```diff
@@ -108,16 +109,28 @@ func BuildLabelSelector(labels map[string]string) metav1.LabelSelector {
+
+	if prefix == "" {
+		return imageName
 	}
```
This will happen if an image is not registered in the list of `Repositories` for the given Solution version. I'd rather fail in the Operator, rather than rely on the fact that _it could work, maybe_.

alexandre-allard replied:
Yes, wasn't sure how to handle this.
Will change it and make it fail.

alexandre-allard replied:
done

slaperche-scality commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -0,0 +1,153 @@
+	Repositories map[string][]Repository `yaml:"repositories" validate:"required,ne=0"`
+}
+
+var operatorConfig OperatorConfig
```
Kept the approach with a global (but made it private to limit the risk) because we couldn't find a way to thread something from the main to the reconciliation loop (the loop has access to a `reconcile.Reconciler` and the only thing we get when creating it is the `manager.Manager`)

If you have a better suggestion, we're open for it.

slaperche-scality replied:
After discussing with @benoit-a I've implemented [this](https://github.com/scality/metalk8s/compare/bugfix/fix-example-solution...user/slaperche/fix-example-solution?expand=1) which allows us to get rid of the global config and thread the config variable from the main to the reconciler.

But this require some modification in the code generated by the SDK, thus I don't know if that's OK for you @gdemonet / @NicolasT or if you prefer the "private global" approach?
Or if you have a 3rd way :slightly_smiling_face: ?

alexandre-allard replied:
I think your Solution is better and we can change the code generated, it's not a problem, but I'm not a Go expert, so waiting for feedback from them.

slaperche-scality replied:
Yeah it's not a problem, except that it may make the upgrade of the SDK a bit more tedious (but  it's already a very manual process).

alexandre-allard replied:
I've cherry-picked your commit.

slaperche-scality committed:
examples: no longer use a singleton for the configuration

Instead of using a singleton pattern for the Operator
configuration, we now pass the configuration through
the manager to make it accessible in controllers.

Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

alexandre-allard committed:
examples: fail if container image is not found in cfg

slaperche-scality commented on examples/metalk8s-solution-example/operator/pkg/config/config.go:
```diff
@@ -80,15 +78,15 @@ func (opConfig *OperatorConfig) validate() error {
 		for _, e := range err.(validator.ValidationErrors) {
 			errorList = append(errorList, getErrorMessage(e, ""))
 		}
 	}
```
Minor: could use the short form (like l.135 of `main.go`)

```go
	if err := validate.Struct(config); err != nil {
		for _, e := range err.(validator.ValidationErrors) {
			errorList = append(errorList, getErrorMessage(e, ""))
		}
	}
```

alexandre-allard committed:
salt: change operator config format

In the operator configuration, we know list the
repositories endpoints & images by solution version.
We need this to tell the Operator where to fetch
the images for a specific solution version.

alexandre-allard committed:
examples: add a default config for example-solution

This default configuration allow action on apiGroups
example-solution.metalk8s.scality.com for the Operator
to be able to monitor CRs

alexandre-allard committed:
examples: add makefile rule to build example solution config

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/fix-example-solution` into
`development/2.4`
* [w/2.5/bugfix/fix-example-solution](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/fix-example-solution) will be merged into `development/2.5` 
* [w/2.6/bugfix/fix-example-solution](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/fix-example-solution) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet commented on examples/metalk8s-solution-example/config.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: solutions.metalk8s.scality.com/v1alpha1
+kind: SolutionConfig
```
Would be cool to show everything we support in this file. Next PR!

alexandre-allard replied:
Yes, good idea.

gdemonet reviewed (approved):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
